packages/rich-text-structure/src/insert.js

```diff
@@ -0,0 +1,28 @@
+/**
+ * Inserts the second given record into the first.
+ * The value in between the start and end indices will be removed.
+ * If no start index or end index is provided, the record's selection will be
+ * used.
+ *
+ * @param {Object} record         Record to modify.
+ * @param {string} recordToInsert Record to insert.
+ * @param {number} startIndex     Start index.
+ * @param {number} endIndex       End index.
+ *
+ * @return {Object} A new record with the record inserted.
+ */
+export function insert(
+	{ formats, text, start, end },
+	recordToInsert,
+	startIndex = start,
+	endIndex = end
+) {
+	const index = startIndex + recordToInsert.text.length;
+
+	return {
+		formats: formats.slice( 0, startIndex ).concat( recordToInsert.formats, formats.slice( endIndex ) ),
```

Comment:
also known as…

```js
formats: [ ...formats ].splice( startIndex, endIndex - startIndex, ...recordToInsert.formats ),
```

Reply:
Yeah. This actually used to be a splice before but without the cloning. I guess `splice` is faster? If speed does not matter, what's the easiest to read? I like that `slice` is used for both `text` and `formats`.

Reply:
it'd be hard to make a sound argument on speed because of how complicated the optimization could end up here. easiest to read? subjective call - I was only trying to point out that the operation you appear to be trying to do "insert things inside another list" already has a core JavaScript function which was designed to do that very thing; in this case `splice` is a stronger semantic match.

it's a shame we don't have string splicing but if I were to guess this is a leaky abstraction from how the original JS engine handled string vs. array data structures, or one suggestion from Google suggested it has to do with the fact that arrays are mutable while strings aren't… 🤷‍♂️

Reply:
Hm, also, we cannot spread a sparse array because every unset index will be set as undefined. `[,]` is not the same as `[...[,]]`.

Reply:
> `[,]` is not the same as `[...[,]]`

that stinks…I confirmed that it works with `slice()` at least. I take back my comment!